lib/elCanvas2.js

```diff
@@ -0,0 +1,807 @@
+			this.objects.push(props);
+			return props;
+		},
+		/**
+		 * draw all my objects.  the "update" function on an instance is a bound version of this
+		 * so it can be passed as a callback without losing context.
+		 */
+		__update: function () {
+			if (this.firstFrame) {
+				_util.addParents(this.objects);
+				this.sortByZIndex();
+			}
+			this.ctx.clearRect(0, 0, 9999, 9999);
+			this.objects.forEach(this.draw.bind(this));
+			this.firstFrame = false;
+		},
+		/**
+		 * sort all my objects so higher z-index come later
+		 */
+		sortByZIndex: function () {
+			this.objects.sort(function (a, b) {
+				return a.zIndex - b.zIndex;
+			});
+		},
+		/**
+		 * remove an object, or array of objects, from my list.
+		 * Will also remove any children of that object, so no orphans
+		 */
+		remove: function (objOrArray) {
+			if (Array.isArray(objOrArray)) {
+				objOrArray.forEach(this.remove.bind(this));
+				return;
+			}
+			var obj = objOrArray;
+			if (obj.childNodes) {
+				this.remove(obj.childNodes);
+				obj.childNodes = null;
+			}
+			var idx = this.objects.indexOf(obj);
+			if (idx > -1) {
+				this.objects.splice(idx, 1);
+			}
+		},
+		addEventListener: function (event, callback) {
+			var target = this.container;
+			var el = this;
+			target.addEventListener(event, function (e) {
+			 var  style = target.currentStyle || window.getComputedStyle(target, null),
+					 borderLeftWidth = parseInt(style['borderLeftWidth'], 10),
+					 borderTopWidth = parseInt(style['borderTopWidth'], 10),
+					 rect = target.getBoundingClientRect(),
+					 scaleX = rect.width / el.width, scaleY = rect.height / el.height,
+					 offsetX = (e.clientX - borderLeftWidth - rect.left) / scaleX,
+					 offsetY = (e.clientY - borderTopWidth - rect.top) / scaleY;
+				 e.canvasX = offsetX, e.canvasY = offsetY;
+				 // the benefit of canvasX, canvasY is they're accurate to position with taking account canvas scaling
+				 callback(e);
+			});
+		},
+		hitTest: function (mousePosition, callback) {
+			var found = this.objects.find(function (object) {
+				return object.interactable && _util.fullyInside(mousePosition, object);
+			});
+			found && callback && callback(found);
+		},
+		addResizeListener: function (callback) {
+			var el = this;
+			var resizeListen = _util.debounce(callback);
+			window.addEventListener('resize', function () {
+				// update canvas sizing
+				el.container.width =  el.container.parentNode.offsetWidth;
+				el.container.height = el.container.parentNode.offsetHeight
+				el.width = el.container.width
+				el.height = el.container.height
+				resizeListen('debounced');
+			}, false);
+		}
+	};
+
+	// Add functions for all object types
+	objectTypes.forEach(function (objectType) {
+		elCanvas.prototype[objectType] = function (suppliedProps) {
+			suppliedProps = _util.mergeDefaultProperties(suppliedProps, objectType);
+			return this.addObject(suppliedProps);
```

Comment:
this.addObject and anything in the this scope isn't accessible anymore, this returns a blank object causing nothing to run